### ISSUE TITLE
Off-by-one additional unexpected retry attempt

### ISSUE THREAD

**Issue by sebgl** [failsafe-go/failsafe-go#22]
I _think_ the retry logic does one more retry than configured.

See the following unit test that succeeds but shouldn't, which expects a function to be attempted 4 times (3 retries), but is actually called 5 times:

```go
func TestMaxAttempts(t *testing.T) {
	policy := retrypolicy.Builder[any]().
		WithMaxAttempts(4). // equivalent to WithMaxRetries(3)
		WithDelay(time.Duration(0)).
		HandleIf(func(_ any, err error) bool {
			return err != nil
		}).
		ReturnLastFailure().
		Build()

	called := 0
	f := func() error {
		called++
		if called == 5 {
			return nil
		}
		return fmt.Errorf("some error")
	}

	err := failsafe.Run(f, policy)
	require.NoError(t, err)
	require.Equal(t, 5, called)
}

```

**Comment by jhalterman** [comment]
This looks as expected. To get 4 attempts, you'd use `WithMaxAttempts(4)` which would give you an initial attempt + 3 retries, and would be equivelent to `WithMaxRetries(3)`.

**Comment by sebgl** [comment]
Sorry I just realized I pasted the wrong sample, I fixed the first post.
I'm using `WithMaxAttempts(4)`, which as I understand is equivalent to `WithMaxRetries(3) `: this calls my function `f` 5 times in total. Rather than 4.

**Comment by jhalterman** [comment]
This turned out to be a lame problem with `ReturnLastFailure`, which didn't have good enough test coverage. If I tweak your test case, it passes as I think you expect now:

```go
func TestMaxAttempts(t *testing.T) {
	policy := retrypolicy.Builder[any]().
		WithMaxAttempts(4). // equivalent to WithMaxRetries(3)
		WithDelay(time.Duration(0)).
		HandleIf(func(_ any, err error) bool {
			return err != nil
		}).
		ReturnLastFailure().
		Build()

	called := 0
	f := func() error {
		called++
		if called == 4 {
			return nil
		}
		return fmt.Errorf("some error")
	}

	err := failsafe.Run(f, policy)
	require.NoError(t, err)
	require.Equal(t, 4, called)
}
```

Let me know if this looks good and I can tag a release.

**Comment by sebgl** [comment]
Thanks, this looks great!

**Comment by jhalterman** [comment]
This was released in 0.4.1.
